File: frontend/mobile/modules/sentinel-collector/android/src/test/java/com/sentinel/collector/BufferPressureTest.kt
package com.sentinel.collector

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The backlog has to read as what it is.
 *
 * "400 buffered" and "12 buffered" look like the same kind of fact — a number
 * that went up. They are not: the first is a buffer at its cap discarding the
 * oldest sample on every new one, and the second is a delay that will resolve
 * itself. The real incident this comes from was read as healthy.
 */
class BufferPressureTest {

    private val capacity = CollectorConfig.BUFFER_SIZE

    @Test
    fun `an empty buffer with nothing lost says nothing at all`() {
        assertNull(BufferPressure.describe(buffered = 0, capacity = capacity, dropped = 0))
    }

    @Test
    fun `an ordinary backlog is just a count`() {
        assertEquals("12 buffered", BufferPressure.describe(12, capacity, dropped = 0))
    }

    @Test
    fun `a full buffer says it is losing data, not just how much it holds`() {
        val text = BufferPressure.describe(capacity, capacity, dropped = 57)

        assertTrue(text!!, text.contains("full"))
        assertTrue(text, text.contains("dropping oldest"))
        assertTrue(text, text.contains("57"))
    }

    @Test
    fun `the cap is named before it is reached, not only after`() {
        // The warning is worth having while there is still something to do
        // about it — the exemption to ask for, the phone to wake.
        val text = BufferPressure.describe((capacity * 0.95).toInt(), capacity, dropped = 0)

        assertTrue(text!!, text.contains("nearly full"))
        assertTrue(text, text.contains("$capacity"))
    }

    @Test
    fun `approaching is not the same as arriving`() {
        val text = BufferPressure.describe((capacity * 0.5).toInt(), capacity, dropped = 0)

        assertEquals("200 buffered", text)
    }

    @Test
    fun `loss is still reported once the backlog has drained`() {
        // A buffer that empties looks like a full recovery. The samples that
        // fell off the front are still gone, and the charts will have a hole
        // in them that nothing else on this screen explains.
        val text = BufferPressure.describe(buffered = 0, capacity = capacity, dropped = 57)

        assertTrue(text!!, text.contains("57 samples lost"))
    }

    @Test
    fun `one lost sample is not reported as one samples`() {
        val text = BufferPressure.describe(buffered = 0, capacity = capacity, dropped = 1)

        assertEquals("1 sample lost while the buffer was full", text)
    }
}

File: frontend/mobile/src/lib/__tests__/bufferPressure.test.ts
import assert from "node:assert/strict"
import { describe, it } from "node:test"

import { describeBuffer } from "../bufferPressure.ts"

/**
 * The distinction being guarded: a backlog is a delay and will resolve itself;
 * a buffer at its cap is discarding the oldest sample on every new one and the
 * history behind it is gone. A plain count renders both as a number going up,
 * which is how a real phone said "400 buffered" for hours and was read as fine.
 */
describe("describeBuffer", () => {
  const capacity = 400

  it("says nothing is waiting rather than reporting a zero", () => {
    const { text, losing } = describeBuffer({
      bufferedSamples: 0,
      bufferCapacity: capacity,
      droppedSamples: 0,
    })
    assert.equal(text, "nothing waiting")
    assert.equal(losing, false)
  })

  it("reports an ordinary backlog as a plain count", () => {
    assert.deepEqual(
      describeBuffer({ bufferedSamples: 12, bufferCapacity: capacity, droppedSamples: 0 }),
      { text: "12 samples", losing: false },
    )
  })

  it("does not pluralise a single sample", () => {
    assert.equal(
      describeBuffer({ bufferedSamples: 1, bufferCapacity: capacity, droppedSamples: 0 }).text,
      "1 sample",
    )
  })

  it("names the cap before it is reached", () => {
    // While there is still something to do about it — grant the battery
    // exemption, wake the phone, get it back on a network.
    const { text, losing } = describeBuffer({
      bufferedSamples: 380,
      bufferCapacity: capacity,
      droppedSamples: 0,
    })
    assert.equal(text, "380 of 400 — nearly full")
    assert.equal(losing, false) // approaching is not arriving
  })

  it("says a full buffer is dropping, not just how much it holds", () => {
    const { text, losing } = describeBuffer({
      bufferedSamples: 400,
      bufferCapacity: capacity,
      droppedSamples: 57,
    })
    assert.equal(text, "400 — full, dropping oldest · 57 lost")
    assert.equal(losing, true)
  })

  it("keeps reporting the loss once the backlog has drained", () => {
    // The buffer emptying looks like a full recovery. The samples that fell
    // off the front are still gone, and the charts will have a hole in them
    // that nothing else on the screen explains.
    assert.deepEqual(
      describeBuffer({ bufferedSamples: 0, bufferCapacity: capacity, droppedSamples: 57 }),
      { text: "57 lost while full", losing: true },
    )
    assert.deepEqual(
      describeBuffer({ bufferedSamples: 8, bufferCapacity: capacity, droppedSamples: 57 }),
      { text: "8 · 57 lost", losing: true },
    )
  })

  it("never claims a capacity on a device that has no collector", () => {
    // UNSUPPORTED_STATUS reports a zero capacity rather than 400, and "0 of 0"
    // would be a ring buffer that does not exist describing how full it is.
    assert.equal(
      describeBuffer({ bufferedSamples: 0, bufferCapacity: 0, droppedSamples: 0 }).text,
      "nothing waiting",
    )
  })
})
